src/merger/merger.cpp

```diff
@@ -0,0 +1,968 @@
+#include "extractor/tarjan_scc.hpp"
+
+#include "guidance/files.hpp"
+#include "guidance/segregated_intersection_classification.hpp"
+
+#include "merger/merger.hpp"
+
+#include "util/log.hpp"
+#include "util/timing_util.hpp"
+
+#include <osmium/handler/node_locations_for_ways.hpp>
+#include <osmium/index/map/flex_mem.hpp>
+#include <osmium/io/any_input.hpp>
+#include <osmium/visitor.hpp>
+
+#if TBB_VERSION_MAJOR == 2020
+#include <tbb/global_control.h>
+#else
+#include <tbb/task_scheduler_init.h>
+#endif
+#include <tbb/pipeline.h>
+
+namespace osrm
+{
+namespace merger
+{
+
+namespace
+{
+void classNamesUnion(std::vector<std::string> &accumulator,
+                     const std::vector<std::string> &class_names)
+{
+    for (const auto &name : class_names)
+    {
+        auto iter = std::find(accumulator.begin(), accumulator.end(), name);
+        if (iter == accumulator.end())
+        {
+            accumulator.push_back(name);
+        }
+    }
+}
+
+void excludeableClassesUnion(std::set<std::set<std::string>> &accumulator,
+                             const std::vector<std::vector<std::string>> &excludable_classes)
+{
+    for (const auto &combination : excludable_classes)
+    {
+        std::set<std::string> combination_set(combination.begin(), combination.end());
+        auto iter = accumulator.find(combination_set);
+        if (iter == accumulator.end())
+        {
+            accumulator.insert(combination_set);
+        }
+    }
+}
+
+// Converts the class name map into a fixed mapping of index to name
+void SetClassNames(const std::vector<std::string> &class_names,
+                   extractor::ExtractorCallbacks::ClassesMap &classes_map,
+                   extractor::ProfileProperties &profile_properties)
+{
+    // if we get a list of class names we can validate if we set invalid classes
+    // and add classes that were never reference
+    if (!class_names.empty())
+    {
+        // add class names that were never used explicitly on a way
+        // this makes sure we can correctly validate unkown class names later
+        for (const auto &name : class_names)
+        {
+            if (!extractor::isValidClassName(name))
+            {
+                throw util::exception("Invalid class name " + name + " only [a-Z0-9] allowed.");
+            }
+
+            auto iter = classes_map.find(name);
+            if (iter == classes_map.end())
+            {
+                auto index = classes_map.size();
+                if (index > extractor::MAX_CLASS_INDEX)
+                {
+                    throw util::exception("Maximum number of classes is " +
+                                          std::to_string(extractor::MAX_CLASS_INDEX + 1));
+                }
+
+                classes_map[name] = extractor::getClassData(index);
+            }
+        }
+
+        // check if class names are only from the list supplied by the user
+        for (const auto &pair : classes_map)
```

Comment:
This if using C++17: you can iterate like  `for (const auto& [a, b] : classes_map)`. In this way you can give more intuitive names to the variables.